File: mobile/src/types/budget.ts
export interface BudgetCategoryInput {
  category_id: string;
  allocated_amount: number;
}

export interface BudgetCreate {
  name: string;
  total_amount: number;
  period?: 'WEEKLY' | 'MONTHLY' | 'YEARLY';
  start_date: string;
  end_date: string;
  categories?: BudgetCategoryInput[];
}

export interface BudgetCategoryStatus {
  category_id: string;
  category_name: string;
  category_icon?: string | null;
  category_color?: string | null;
  allocated_amount: number | string;
  spent_amount: number | string;
  remaining_amount: number | string;
  utilization_pct: number;
  status: 'NORMAL' | 'WARNING' | 'EXCEEDED';
  warning_message?: string | null;
}

export interface Budget {
  id: string;
  user_id: string;
  name: string;
  total_amount: number | string;
  period: string;
  start_date: string;
  end_date: string;
  is_active: boolean;
  created_at: string;
  total_spent: number | string;
  total_remaining: number | string;
  total_utilization_pct: number;
  overall_status: 'NORMAL' | 'WARNING' | 'EXCEEDED';
  categories: BudgetCategoryStatus[];
}

File: backend/app/schemas/budget.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field

class BudgetCategoryInput(BaseModel):
    category_id: UUID
    allocated_amount: Decimal = Field(..., gt=0, decimal_places=2, max_digits=12)

class BudgetCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    total_amount: Decimal = Field(..., gt=0, decimal_places=2, max_digits=12)
    period: str = Field("MONTHLY", pattern="^(WEEKLY|MONTHLY|YEARLY)$")
    start_date: date
    end_date: date
    categories: List[BudgetCategoryInput] = []

class BudgetUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    total_amount: Optional[Decimal] = Field(None, gt=0, decimal_places=2, max_digits=12)
    period: Optional[str] = Field(None, pattern="^(WEEKLY|MONTHLY|YEARLY)$")
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    is_active: Optional[bool] = None
    categories: Optional[List[BudgetCategoryInput]] = None

class BudgetCategoryStatus(BaseModel):
    category_id: UUID
    category_name: str
    category_icon: Optional[str] = None
    category_color: Optional[str] = None
    allocated_amount: Decimal
    spent_amount: Decimal
    remaining_amount: Decimal
    utilization_pct: float
    status: str  # NORMAL, WARNING, EXCEEDED
    warning_message: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)

class BudgetResponse(BaseModel):
    id: UUID
    user_id: UUID
    name: str
    total_amount: Decimal
    period: str
    start_date: date
    end_date: date
    is_active: bool
    created_at: datetime
    total_spent: Decimal = Decimal("0.00")
    total_remaining: Decimal = Decimal("0.00")
    total_utilization_pct: float = 0.0
    overall_status: str = "NORMAL"
    categories: List[BudgetCategoryStatus] = []

    model_config = ConfigDict(from_attributes=True)
